// allure-cypress/dist/reporter.mjs
import { readFileSync } from 'node:fs';
import { getSuitesLabels, extractMetadataFromString, AllureTest, Stage, LabelName } from 'allure-js-commons/new';
import { AllureNodeRuntime, FileSystemAllureWriter } from 'allure-js-commons/new/node';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol */


function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var MessageType;
(function (MessageType) {
    MessageType["TEST_STARTED"] = "TEST_STARTED";
    MessageType["TEST_ENDED"] = "TEST_ENDED";
    MessageType["STEP_STARTED"] = "STEP_STARTED";
    MessageType["STEP_ENDED"] = "STEP_ENDED";
    MessageType["METADATA"] = "METADATA";
    MessageType["SCREENSHOT"] = "SCREENSHOT";
})(MessageType || (MessageType = {}));

const startAllureTest = (runtime, message) => {
    const suiteLabels = getSuitesLabels(message.specPath.slice(0, -1));
    const testTitle = message.specPath[message.specPath.length - 1];
    const titleMetadata = extractMetadataFromString(testTitle);
    const currentTest = new AllureTest(runtime, message.start);
    currentTest.name = titleMetadata.cleanTitle;
    currentTest.fullName = `${message.filename}#${message.specPath.join(" ")}`;
    currentTest.stage = Stage.RUNNING;
    currentTest.addLabel(LabelName.LANGUAGE, "javascript");
    currentTest.addLabel(LabelName.FRAMEWORK, "cypress");
    suiteLabels.forEach((label) => {
        currentTest.addLabel(label.name, label.value);
    });
    titleMetadata.labels.forEach((label) => {
        currentTest.addLabel(label.name, label.value);
    });
    return currentTest;
};
const allureCypress = (on, config) => {
    const runtime = new AllureNodeRuntime({
        writer: new FileSystemAllureWriter({
            resultsDir: (config === null || config === void 0 ? void 0 : config.resultsDir) || "./allure-results",
        }),
    });
    const currentSteps = [];
    on("task", {
        allureReportTest: ({ startMessage, endMessage, messages }) => {
            const currentTest = startAllureTest(runtime, startMessage);
            messages.forEach(({ type, payload }) => {
                var _a;
                if (type === MessageType.STEP_STARTED) {
                    const currentStep = currentSteps[currentSteps.length - 1];
                    const newStep = (currentStep || currentTest).startStep(payload.name, payload.start);
                    currentSteps.push(newStep);
                    return;
                }
                if (type === MessageType.STEP_ENDED) {
                    const currentStep = currentSteps.pop();
                    currentStep.status = payload.status;
                    currentStep.statusDetails = payload.statusDetails;
                    currentStep.stage = payload.stage;
                    currentStep.endStep(payload.stop);
                    return;
                }
                if (type === MessageType.SCREENSHOT) {
                    const currentStep = currentSteps[currentSteps.length - 1];
                    const attachmentName = payload.name;
                    const screenshotBody = readFileSync(payload.path);
                    const screenshotName = runtime.writeAttachment(screenshotBody, "image/png");
                    (currentStep || currentTest).addAttachment(attachmentName, "image/png", screenshotName);
                    return;
                }
                if (type === MessageType.METADATA) {
                    const { parameter = [], links = [], attachments } = payload, metadata = __rest(payload, ["parameter", "links", "attachments"]);
                    const currentStep = currentSteps[currentSteps.length - 1];
                    parameter.forEach(({ name, value, excluded, mode }) => {
                        currentTest.parameter(name, value, {
                            excluded,
                            mode,
                        });
                    });
                    attachments === null || attachments === void 0 ? void 0 : attachments.forEach((attachment) => {
                        const attachmentName = runtime.writeAttachment(attachment.content, attachment.type, attachment.encoding);
                        (currentStep || currentTest).addAttachment(attachment.name, attachment.type, attachmentName);
                    });
                    if (!((_a = config === null || config === void 0 ? void 0 : config.links) === null || _a === void 0 ? void 0 : _a.length)) {
                        currentTest.applyMetadata(Object.assign(Object.assign({}, metadata), { links }));
                        return;
                    }
                    const formattedLinks = links.map((link) => {
                        var _a, _b;
                        const matcher = (_b = (_a = config === null || config === void 0 ? void 0 : config.links) === null || _a === void 0 ? void 0 : _a.find) === null || _b === void 0 ? void 0 : _b.call(_a, (item) => item.type === link.type);
                        if (!matcher || link.url.startsWith("http")) {
                            return link;
                        }
                        const url = matcher.urlTemplate.replace("%s", link.url);
                        const name = link.name || link.url;
                        return Object.assign(Object.assign({}, link), { name,
                            url });
                    });
                    currentTest.applyMetadata(Object.assign(Object.assign({}, metadata), { links: formattedLinks }));
                    return;
                }
            });
            currentTest.stage = endMessage.stage;
            currentTest.status = endMessage.status;
            currentTest.statusDetails = endMessage.statusDetails;
            currentTest.calculateHistoryId();
            currentTest.endTest(endMessage.stop);
            return null;
        },
    });
};

export { allureCypress };
